### PR TITLE
[SPARK-36227][SQL][3.2] Remove TimestampNTZ type support in Spark 3.2

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Remove TimestampNTZ type support in the production code of Spark 3.2. 
To archive the goal, this PR adds the check "Utils.isTesting" in the following code branches:
- keyword "timestamp_ntz" and "timestamp_ltz" in parser
- New expressions from https://issues.apache.org/jira/browse/SPARK-35662
- Using java.time.localDateTime as the external type for TimestampNTZType
- `SQLConf.timestampType` which determines the default timestamp type of Spark SQL.

This is to minimize the code difference between the master branch. So that future users won't think TimestampNTZ is already available in Spark 3.2.
The downside is that users can still find TimestampNTZType under package `org.apache.spark.sql.types`. There should be nothing left other than this.
### Why are the changes needed?

As of now, there are some blockers for delivering the TimestampNTZ project in Spark 3.2:

- In the Hive Thrift server, both TimestampType and TimestampNTZType are mapped to the same timestamp type, which can cause confusion for users.
- For the Parquet data source, the new written TimestampNTZType Parquet columns will be read as TimestampType in old Spark releases. Also, we need to decide the merge schema for files mixed with TimestampType and TimestampNTZ type.
- The type coercion rules for TimestampNTZType are incomplete. For example, what should the data type of the in clause "IN(Timestamp'2020-01-01 00:00:00', TimestampNtz'2020-01-01 00:00:00') be.
- It is tricky to support TimestampNTZType in JSON/CSV data readers. We need to avoid regressions as possible as we can.

There are 10 days left for the expected 3.2 RC date. So, I propose to **release the TimestampNTZ type in Spark 3.3 instead of Spark 3.2**. So that we have enough time to make considerate designs for the issues.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing Unit tests + manual tests from spark-shell to validate the changes are gone.
New functions 
```
spark.sql("select to_timestamp_ntz'2021-01-01 00:00:00'").show()
spark.sql("select to_timestamp_ltz'2021-01-01 00:00:00'").show()
spark.sql("select make_timestamp_ntz(1,1,1,1,1,1)").show()
spark.sql("select make_timestamp_ltz(1,1,1,1,1,1)").show()
spark.sql("select localtimestamp()").show()
```
The SQL configuration `spark.sql.timestampType` should not work in 3.2
```
spark.conf.set("spark.sql.timestampType", "TIMESTAMP_NTZ")
spark.sql("select make_timestamp(1,1,1,1,1,1)").schema
spark.sql("select to_timestamp('2021-01-01 00:00:00')").schema
spark.sql("select timestamp'2021-01-01 00:00:00'").schema
Seq((1, java.sql.Timestamp.valueOf("2021-01-01 00:00:00"))).toDF("i", "ts").write.partitionBy("ts").parquet("/tmp/test")
spark.read.parquet("/tmp/test").schema
```
LocalDateTime is not supported as a built-in external type:
```
Seq(LocalDateTime.now()).toDF()
org.apache.spark.sql.catalyst.expressions.Literal(java.time.LocalDateTime.now())
org.apache.spark.sql.catalyst.expressions.Literal(0L, TimestampNTZType)
```

### CODE REVIEW

**Comment by gengliangwang** [comment]
Marking this PR as WIP so that I can observe the CI test results. I will add new test cases as well.

**Review by dongjoon-hyun — COMMENTED** [review]
Is this a complete list for the removal, @gengliangwang ?

**Comment by HyukjinKwon** [comment]
Why don't we just simply document?

**Comment by HyukjinKwon** [comment]
in fact we didn't even yet document it as an official type in https://spark.apache.org/docs/latest/sql-ref-datatypes.html. If we're worried about being used, I would just explicitly document it there and mention that this is incomplete. Users won't be able to know there's a new type unless they read codes anyway.

**Comment by HyukjinKwon** [comment]
okie, had a short offline discussion. if this is all we need to change (or if the change is not so big), I am fine.

**Comment by gengliangwang** [comment]
@dongjoon-hyun @HyukjinKwon I think this is almost all of them. I am still double-checking.

**Comment by gengliangwang** [comment]
This is ready for review.
cc @dongjoon-hyun @HyukjinKwon @cloud-fan @MaxGekk @beliefer @yaooqinn

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -3975,12 +3976,14 @@ class SQLConf extends Serializable with Logging {
   def ansiEnabled: Boolean = getConf(ANSI_ENABLED)
 
   def timestampType: AtomicType = getConf(TIMESTAMP_TYPE) match {
-    case "TIMESTAMP_LTZ" =>
+    // SPARK-36227: Remove TimestampNTZ type support in Spark 3.2 with minimal code changes.
+    //              The configuration `TIMESTAMP_TYPE` is only effective for testing in Spark 3.2.
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3980]: indentation ?

**Review by beliefer — APPROVED** [review]
LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3980]: Oh, this is on purpose.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala [hunk 2]
```diff
@@ -114,14 +114,6 @@ object Encoders {
    */
   def LOCALDATE: Encoder[java.time.LocalDate] = ExpressionEncoder()
 
-  /**
-   * Creates an encoder that serializes instances of the `java.time.LocalDateTime` class
-   * to the internal representation of nullable Catalyst's TimestampNTZType.
-   *
-   * @since 3.2.0
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:121]: we need update master branch

**Review by yaooqinn — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala [hunk 3]
```diff
@@ -297,10 +297,6 @@ package object dsl {
       /** Creates a new AttributeReference of type timestamp */
       def timestamp: AttributeReference = AttributeReference(s, TimestampType, nullable = true)()
 
-      /** Creates a new AttributeReference of type timestamp without time zone */
-      def timestampNTZ: AttributeReference =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:301]: this is an internal API

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala:121]: Yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala:301]: End users can still access it. Let's remove it anyway

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to branch-3.2.

**Comment by dongjoon-hyun** [comment]
Thank you, @gengliangwang and all.
@gengliangwang . Please proceed to update the master branch.

**Pull request closed without merging**